sky/adaptors/aws.py

```diff
@@ -41,8 +42,23 @@ def session():
     # Reference: https://github.com/boto/boto3/issues/1592
     # However, the session object itself is thread-safe, so we are
     # able to use lru_cache() to cache the session object.
-    with _session_creation_lock:
-        return boto3.session.Session()
+
+    # Retry 5 times by default for potential credential errors,
+    # mentioned in
+    # https://github.com/skypilot-org/skypilot/pull/1988
+    max_attempts = 5
+    attempt = 0
+    err = None
+    while attempt < max_attempts:
+        try:
+            with _session_creation_lock:
+                return boto3.session.Session()
+        except (botocore_exceptions().CredentialRetrievalError,
```

Comment:
Actually, I'm not too sure whether the error reported by the user in #1988 was thrown at session create time. There was no stacktrace. Spot controller log:
```
I 05-29 16:23:00 spot_utils.py:69] === Checking the job status... ===
I 05-29 16:23:01 spot_utils.py:75] Job status: JobStatus.RUNNING
I 05-29 16:23:01 spot_utils.py:78] ==================================
sky.exceptions.ClusterStatusFetchingError: Failed to query AWS cluster 'train-xxxxxxx-16' status: Unable to locate credentials. You can configure credentials by running "aws configure".


Unexpected error occurred: [sky.exceptions.ClusterStatusFetchingError] Failed to query AWS cluster 'train-xxxxxxx-16' status: Unable to locate credentials. You can configure credentials by running "aws configure".

I 05-29 16:23:06 spot_state.py:292] Unexpected error occurred: [sky.exceptions.ClusterStatusFetchingError] Failed to query AWS cluster 'train-xxxxxxx-16' status: Unable to locate credentials. You can configure credentials by running "aws configure".
```

Guessing from the newer stacktrace in #2301, is it possible that NoCredentialsError is thrown not at creation time by at request time?

Another possibility: #2301 happened because the current session/resource handling are not thread-safe.

Reply:
hmmm, if it happens at request time, then I do not have good approaches to handle this (seems impossible to distinguish between a real error or flaky env issue). But this may not be an issue, since we add default retries here: https://github.com/skypilot-org/skypilot/blob/ca2a092a1cc26d26fb49f94a11cfe5f369df0f25/sky/provision/aws/instance.py#L16-L21